### PR TITLE
Fix configuring `astropy` cache location

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

657c91a53e7bf049a7e29e9910bcbdd8453bd3dd (from #16997) introduced a bug to customizing the `astropy` cache directory. The `XDG_CACHE_HOME` environment variable has no effect, instead `XDG_CONFIG_HOME` controls both configuration and cache directories.

The first commit adds a regression test to reveal the bug, the second fixes it.

PS https://www.astropy.org/team does not document who the maintainers for `config` are.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Looks good to me, thanks @eerovaher!

File: astropy/config/paths.py
Comment (saimn) [inline astropy/config/paths.py:31]: Might be useful to specify above that `fallback` can take a limited number of values.

Comment (neutrinoceros) [reply astropy/config/paths.py:31]: yes. We could also expose the environment variable directly as a separate argument, since it's a private function, there's no pressure to preserve its API.

**Comment by eerovaher** [comment]
I've figured out why the double test CI job failed, but I don't have a good solution how to get it to work. The double test job uses the `astropy` test runner, whereas the other jobs run `pytest` directly: https://github.com/astropy/astropy/blob/2e7cf84c32938b36b682ab42153348ffdb2d9f12/tox.ini#L125
The `astropy` test runner configures the cache directory using `set_temp_cache()`: https://github.com/astropy/astropy/blob/2e7cf84c32938b36b682ab42153348ffdb2d9f12/astropy/tests/runner.py#L253
`set_temp_cache()` is supposed to have priority over environment variables, which fully explains why changing the environment variables in the new test has no effect in that one CI job.

The best solution would be to skip the new test if and only if the tests are run using the `astropy` test runner, but I don't know how to do that. A simple solution would be to use `set_temp_cache()` and `set_temp_config()` in the test, but I don't like the idea of adding these extra steps to the test just because of the `astropy` test runner.

**Review by pllim — DISMISSED** [review]
Good catch. Thanks!

Original `config` designer was probably `mdboom` who left some time ago. 🙈

**Comment by pllim** [comment]
Your tests seem to fail in the "double" run, so that needs to be fixed before this can be merged: https://github.com/astropy/astropy/actions/runs/12203451342/job/34046435787?pr=17514

**Comment by eerovaher** [comment]
> Your tests seem to fail in the "double" run, so that needs to be fixed before this can be merged: https://github.com/astropy/astropy/actions/runs/12203451342/job/34046435787?pr=17514

I find it quite disturbing that you have approved this pull request when it is clear that [you have not read the discussion](https://github.com/astropy/astropy/pull/17514#issuecomment-2528372062) at all.

**Comment by mhvk** [comment]
Like many (certainly like me), @pllim probably looks at the actual changes and approves those, and then realizes a test is broken. That that was discussed further above just goes to show that the github way of organizing things is far from optimal -- reading longer discussions is extremely tedious since there is no organization, no way to resolve pieces of the main thread, etc. 

Anyway, @pllim, know we know you are the one that keeps this ship going! Thanks!!

**Comment by pllim** [comment]
Yes, indeed I jumped to the diff first. I do not have time to always read all the discussions, especially since now I am trying to play catch up after being away. But that is neither here nor there. I should probably dismiss my approval for now. Thanks, all.

**Comment by pllim** [comment]
Re: double trouble -- A workaround is to set a `IS_DOUBLE` for the "double" job and have a test skip on that condition. Maybe here:

https://github.com/astropy/astropy/blob/3ca1a101fdea9e88257811abfb51a5435773a2de/tox.ini#L15

Or maybe like this:

https://github-actions-workflows.openastronomy.org/en/stable/tox.html#setenv

Not pretty but it would disappear when we remove the test runner.

**Comment by neutrinoceros** [comment]
I just merged #17529 and set its backport to auto-merge.

File: astropy/config/paths.py
Comment (eerovaher) [reply astropy/config/paths.py:31]: A good way to do that would be to change the type annotation of the `fallback` parameter from `str` to `Literal["cache", "config"]`, but it would also be a good idea to update the annotation of the `cls` parameter from `type` to `type[_SetTempPath]` and there is more room for improvement elsewhere in this file. My preference would be to only make minimal required changes here to fix the bug and to open a separate pull request for the annotations.

Comment (neutrinoceros) [reply astropy/config/paths.py:32]: Sounds like a plan to me. Final question before I merge: do you want to do the follow up yourself too ? (either way is fine)

**Review by neutrinoceros — APPROVED** [review]

**Comment by neutrinoceros** [comment]
Approving now but leaving time for @eerovaher to answer [the one hanging thread](https://github.com/astropy/astropy/pull/17514#discussion_r1873906626) before I merge

File: astropy/config/paths.py
Comment (eerovaher) [reply astropy/config/paths.py:32]: I wouldn't mind updating the annotations myself.

Comment (neutrinoceros) [reply astropy/config/paths.py:32]: alright then :shipit:

**Pull request merged**
